pages/api/user/index.ts

```diff
@@ -0,0 +1,180 @@
+  }
+
+  let oldUserDetails: UserDetailsFromRequest;
+  try {
+    oldUserDetails = await getUserFromJwt(oldUserToken);
+  } catch (error) {
+    return res
+      .status(HttpStatus.UNAUTHORIZED)
+      .json(createJsonResponse({}, { type: ERROR, message: `Could not validate and decode 'old_token'.` }));
+  }
+
+  const oldUid = oldUserDetails.uid;
+  if (!oldUserDetails.isAnonymous) {
+    return res
+      .status(HttpStatus.BAD_REQUEST)
+      .json(createJsonResponse({}, { type: ERROR, message: `UID ${oldUid} is already linked` }));
+  }
+
+  const result = await linkAccount(oldUid, currentUid);
+
+  switch (result) {
+    case LinkAccountResult.OLD_USER_DELETED:
+      return res.status(HttpStatus.OK).json(
+        createJsonResponse(
+          {},
+          {
+            type: INFORMATION,
+            message: `Old UID ${oldUid} is no longer in database. New user ${currentUid} has been linked.`,
+          },
+        ),
+      );
+
+    case LinkAccountResult.NEW_USER_HAS_DATA:
+      return res
+        .status(HttpStatus.CONFLICT)
+        .json(createJsonResponse({}, { type: ERROR, message: 'New UID already has user data and cannot be linked.' }));
+
+    case LinkAccountResult.NEW_USER_LINKED:
+      return res
+        .status(HttpStatus.OK)
+        .json(
+          createJsonResponse(
+            {},
+            { type: SUCCESS, message: `Old UID ${oldUid} has been replaced with new UID ${currentUid}.` },
+          ),
+        );
+  }
+}
+
+async function handleDelete(uid: string, _req: NextApiRequest, res: NextApiResponse<ApiResponse<EmptyPayload>>) {
+  // UID is a primary key, so the count will either be 0 (UID does not exist in database) or 1 (successful delete).
+  const user = await getUserIfExists(uid);
+  if (!user) {
+    return res
+      .status(HttpStatus.NOT_FOUND)
+      .json(createJsonResponse({}, { type: ERROR, message: `Could not find UID ${uid}.` }));
+  }
+
+  await prisma.application.deleteMany({ where: { userId: uid } });
+  await prisma.user.delete({ where: { uid } });
+
+  return res
+    .status(HttpStatus.OK)
+    .json(createJsonResponse({}, { type: SUCCESS, message: `Deleted UID ${uid} and their data.` }));
+}
+
+async function linkAccount(oldUid: string, newUid: string): Promise<LinkAccountResult> {
+  const newUser = await getUserIfExists(newUid);
+  const oldUser = await getUserIfExists(oldUid);
+
+  // If old user has already been removed from the database. This can happen if the API was called twice or if the new user
+  // creation process failed and the API was retried.
+  if (!oldUser) {
+    if (!newUser) {
+      await prisma.user.create({ data: { uid: newUid } });
+    }
+
+    return LinkAccountResult.OLD_USER_DELETED;
+  }
+
+  // Handle situation where new user exists already with data.
+  if (newUser && hasUserData(newUser)) {
+    return LinkAccountResult.NEW_USER_HAS_DATA;
+  }
+
+  // New user exists but has no data (linking should continue as if new user does not exist).
```

Comment:
When will this case happen? I think it's a bit confusing to think about it as `oldUser` & `newUser` compared to `anonUser` and `verifiedUser` because the only time we should `linkAccount` is between an `anonUser` and `verifiedUser`?

Once `verifiedUser` exists, any future attempts to link an `anonUser` with `verifiedUser` shouldn't work?

Reply:
This could happen in the following flow:

1. User opens on browser X (anon1)
2. User signs in with GitHub on X (verified1): there is still no data
3. User opens up on browser Y (anon2)
4. User adds data on browser Y (anon2)
5. User remembers about the GitHub accounts.
6. User signs in with Github on Y (verified1)

With my flow, on step 6 verified1 gets the data created by anon2.

We could deal with it the way you mentioned, but I wanted to improve UX and not reject an attempt to link if the verified doesn't have data.

Reply:
I don't think this makes sense because once a `verifiedUser` is created, we cannot assume that future log ins of an `anonUser` to a `verifiedUser` means that `anonUser` is the `verifiedUser`.

Actually even the first time is a bit iffy isn't it? Should we never take anon user's data when we log in?

Suppose you're tracking your data anonymously on someone else's browser but you forget to clear cache, then anyone else can log in and take your data?

Reply:
Not possible to 'take' someone else's data. If you forgot to clear cache, there's a log in button. To log in with GitHub, you would be clicking something like 'Link account', similar to other web apps out there. If you can reach the 'link account' while logged in (aka with the anon token still in ur browser), then you get that data.

Reply:
Resolved: we will never allow relinking.